File: src/main/java/com/micronaut/bug/controller/SecurityContextAspect.java
package com.micronaut.bug.controller;

import com.micronaut.bug.config.User;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;

import java.lang.annotation.Annotation;

//@Aspect
//@Component
public class SecurityContextAspect {
/*
    @Around("execution(* com.micronaut.bug.controller..*(.., @CurrentUser (*), ..))")
    public Object injectUser(ProceedingJoinPoint joinPoint) {
        return Mono.deferContextual(ctx -> {
            User user = ctx.getOrDefault("current-user", null);

            // Получаем все аргументы метода
            Object[] args = joinPoint.getArgs();
            var signature = (MethodSignature) joinPoint.getSignature();
            Annotation[][] parameterAnnotations = signature.getMethod().getParameterAnnotations();

            // Ищем, какой аргумент помечен @CurrentUser
            for (int i = 0; i < parameterAnnotations.length; i++) {
                var argFound = false;
                for (Annotation annotation : parameterAnnotations[i]) {
                    if (annotation instanceof CurrentUser) {
                        args[i] = user; // Подставляем юзера из контекста
                        argFound = true;
                        break;
                    }
                }
                if (argFound) {
                    break;
                }
            }

            try {
                // Выполняем метод контроллера с подставленным юзером
                Object result = joinPoint.proceed(args);

                // Если метод вернул Mono (а он должен), возвращаем его
                if (result instanceof Mono<?> mono) {
                    return mono;
                }
                return Mono.justOrEmpty(result);
            } catch (Throwable e) {
                return Mono.error(e);
            }
        });
    }*/
}

File: src/main/kotlin/com/micronaut/bug/controller/SecurityContextAspect.kt
package com.micronaut.bug.controller

import com.micronaut.bug.config.User
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.reflect.MethodSignature
import org.springframework.stereotype.Component
//import reactor.core.publisher.Mono

//@Aspect
//@Component
class SecurityContextAspect {
/*

    @Around("execution(* com.micronaut.bug.controller..*(.., @CurrentUser (*), ..))")
    fun injectUser(joinPoint: ProceedingJoinPoint): Any? {
        return Mono.deferContextual { ctx ->
            val user = ctx.getOrDefault<User>("current-user", null)

            // Получаем все аргументы метода
            val args = joinPoint.args
            val signature = joinPoint.signature as MethodSignature
            val parameterAnnotations = signature.method.parameterAnnotations

            // Ищем, какой аргумент помечен @CurrentUser
            for (i in parameterAnnotations.indices) {
                val annotations = parameterAnnotations[i]
                if (annotations.any { it is CurrentUser }) {
                    args[i] = user // Подставляем юзера из контекста
                    break
                }
            }

            try {
                // Выполняем метод контроллера с подставленным юзером
                val result = joinPoint.proceed(args)

                // Если метод вернул Mono, возвращаем его, иначе оборачиваем
                if (result is Mono<*>) {
                    result
                } else {
                    Mono.justOrEmpty(result)
                }
            } catch (e: Throwable) {
                Mono.error<Any>(e)
            }
        }
    }
*/
}
